dartagnan/src/main/java/com/dat3m/dartagnan/expression/integers/IntUnaryExpr.java

```diff
@@ -0,0 +1,61 @@
+package com.dat3m.dartagnan.expression.integers;
+
+import com.dat3m.dartagnan.expression.Expression;
+import com.dat3m.dartagnan.expression.ExpressionVisitor;
+import com.dat3m.dartagnan.expression.base.UnaryExpressionBase;
+import com.dat3m.dartagnan.expression.type.IntegerType;
+
+import java.math.BigInteger;
+
+import static com.google.common.base.Verify.verify;
+
+public class IntUnaryExpr extends UnaryExpressionBase<IntegerType, IntUnaryOp> {
+
+    public IntUnaryExpr(IntUnaryOp operator, Expression operand, IntegerType t) {
+        super(t, operator, operand);
+    }
+
+    @Override
+    public IntLiteral reduce() {
+        if (!(operand.getType() instanceof IntegerType innerType)) {
+            throw new IllegalStateException(String.format("Non-integer operand %s.", operand));
+        }
+        IntLiteral inner = operand.reduce();
+        verify(inner.getType().equals(innerType),
+                "Reduced to wrong type %s instead of %s.", inner.getType(), innerType);
+        BigInteger value = inner.getValue();
+        IntegerType targetType = getType();
+        switch (kind) {
+            case MINUS -> {
+                return new IntLiteral(targetType, value.negate());
+            }
+            case CTLZ -> {
+                if (innerType.isMathematical()) {
+                    throw new UnsupportedOperationException(
+                            String.format("Counting leading zeroes in mathematical integer %s.", inner));
+                }
+                if (value.signum() == -1) {
+                    return new IntLiteral(targetType, BigInteger.ZERO);
+                }
```

Comment:
I don't get this check

Reply:
Negative numbers have their highest bit set, so there are no leading zeroes.